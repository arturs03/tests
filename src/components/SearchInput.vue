<script setup lang="ts">
import { ref } from 'vue'

const props = defineProps<{ list?: string[] }>()

const emit = defineEmits<{
  (e: 'search', item: string): void
  (e: 'selected', item: string): void
}>()

function setPickedItem(e: MouseEvent | KeyboardEvent) {
  const target = e.target as HTMLElement
  if (!target) {
    return
  }

  if (target.className.includes('suggested-list__item')) {
    emit('selected', target.innerHTML)
  }
}

function handleInput(e: KeyboardEvent) {
  const target = e.target as HTMLInputElement
  if (!target) {
    return
  }

  const value = target.value
  debounce(() => emit('search', value))
}

const timeout = ref<number | undefined>(0)
function debounce(callback: () => void) {
  if (timeout.value) {
    clearTimeout(timeout.value)
  }

  timeout.value = setTimeout(() => {
    callback()
    timeout.value = undefined
  }, 400)
}
</script>
<template>
  <div class="search-input">
    <input type="text" @keyup="handleInput" placeholder="Search..." data-test="search-input" />
    <div class="suggested-wrapper">
      <div
        v-show="props.list?.length"
        class="suggested-list"
        @click="setPickedItem"
        @keyup.enter="setPickedItem"
      >
        <span
          v-for="(suggestion, index) in props.list"
          :key="index"
          v-focusable
          class="suggested-list__item"
        >
          {{ suggestion }}
        </span>
      </div>
    </div>
  </div>
</template>
<style lang="scss" scoped>
.search-input {
  max-width: 18.75rem;
  width: 100%;

  input {
    width: 100%;
    border-radius: 0.5rem;
    height: 2rem;
    border: unset;
    padding: 0.25rem 0.5rem;
  }

  .suggested-wrapper {
    position: relative;
  }

  .suggested-list {
    position: absolute;
    top: 0;
    left: 0;
    box-shadow: 5px 5px black;
    border-radius: 0.5rem;
    width: 100%;
    max-height: 18.75rem;
    background-color: white;
    padding: 0.5rem;
    display: flex;
    flex-direction: column;
    gap: 0.5rem;
    color: black;
    overflow-y: scroll;

    &__item {
      list-style: unset;
      border-bottom: 1px solid transparent;

      &:hover {
        cursor: pointer;
        border-bottom: 1px solid black;
      }
    }
  }
}
</style>
