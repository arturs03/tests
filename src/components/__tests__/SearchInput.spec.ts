import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import SearchInput from '@/components//SearchInput.vue'
import { nextTick } from 'vue'

describe('SearchInput', async () => {
  const wrapper = mount(SearchInput)

  beforeEach(() => {
    vi.useFakeTimers()
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  const searchInputEl = wrapper.find('[data-test="search-input"]')

  it('renders component correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('suggested list', async () => {
    const matchList = ['apple', 'apricot', 'almond', 'anise', 'azalea']

    it('emits input value', async () => {
      await searchInputEl.setValue('a')
      await searchInputEl.trigger('keyup', { key: 'a' })

      vi.advanceTimersByTime(400)
      await nextTick()

      expect(wrapper.emitted().search[0]).toEqual(['a'])
    })

    it('re-renders list items from 0 to 5', async () => {
      const oldList = getListItemElements(wrapper)
      expect(oldList.length).toEqual(0)

      await wrapper.setProps({
        list: matchList,
      })
      await nextTick()

      const newList = getListItemElements(wrapper)
      expect(newList).toMatchObject(matchList)
    })

    it('emits selected value "apple"', async () => {
      const firstItem = wrapper.find('.suggested-list__item')
      await firstItem.trigger('click')
      expect(wrapper.emitted().selected[0]).toEqual([matchList[0]])
    })
  })
})

function getListItemElements(wrapper: VueWrapper): string[] {
  return wrapper.findAll('.suggested-list__item').map((item) => item.element.innerHTML)
}
