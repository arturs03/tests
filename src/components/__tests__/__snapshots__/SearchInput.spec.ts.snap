// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchInput > renders component correctly 1`] = `
"<div data-v-54e32457="" class="search-input"><input data-v-54e32457="" type="text" placeholder="Search..." data-test="search-input">
  <div data-v-54e32457="" class="suggested-wrapper">
    <div data-v-54e32457="" class="suggested-list" style="display: none;"></div>
  </div>
</div>"
`;
